File: bindings/nodejs/test/secretbox.test.js
'use strict';

/**
 * `crypto_secretbox` - three categories per D-64/D-65: correctness (round trip - no official
 * vector exists for this construction, same posture as the Rust crate's own tests, D-51),
 * rejection (tamper/wrong key), misuse (wrong-length key, truncated input). Mirrors
 * bindings/python/tests/test_secretbox.py.
 */

const test = require('node:test');
const assert = require('node:assert/strict');
const dstu = require('../js/index.js');

test('seal/open round-trips', () => {
  const key = dstu.secretboxKeygen();
  const sealed = dstu.secretboxSeal(key, Buffer.from('a message worth protecting'));
  assert.deepStrictEqual(dstu.secretboxOpen(key, sealed), Buffer.from('a message worth protecting'));
});

test('seal handles empty message', () => {
  const key = dstu.secretboxKeygen();
  const sealed = dstu.secretboxSeal(key, Buffer.alloc(0));
  assert.deepStrictEqual(dstu.secretboxOpen(key, sealed), Buffer.alloc(0));
});

test('tampered ciphertext is rejected', () => {
  const key = dstu.secretboxKeygen();
  const sealed = Buffer.from(dstu.secretboxSeal(key, Buffer.from('message')));
  sealed[sealed.length - 1] ^= 1;
  assert.throws(() => dstu.secretboxOpen(key, sealed));
});

test('tampered nonce is rejected', () => {
  const key = dstu.secretboxKeygen();
  const sealed = Buffer.from(dstu.secretboxSeal(key, Buffer.from('message')));
  sealed[0] ^= 1; // first byte of the nonce
  assert.throws(() => dstu.secretboxOpen(key, sealed));
});

test('wrong key is rejected', () => {
  const key = dstu.secretboxKeygen();
  const otherKey = dstu.secretboxKeygen();
  const sealed = dstu.secretboxSeal(key, Buffer.from('message'));
  assert.throws(() => dstu.secretboxOpen(otherKey, sealed));
});

test('wrong-length key is rejected', () => {
  assert.throws(() => dstu.secretboxSeal(Buffer.from('too short'), Buffer.from('message')));
});

test('truncated sealed input is rejected', () => {
  const key = dstu.secretboxKeygen();
  assert.throws(() => dstu.secretboxOpen(key, Buffer.from('short')));
});

File: bindings/java/src/test/java/ua/dstucrypto/dstucore/SecretBoxTest.java
package ua.dstucrypto.dstucore;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

/**
 * {@code crypto_secretbox} - three test categories per D-64/D-65: correctness (round trip, no
 * official vector exists for this construction, D-51), rejection (tamper/wrong key), misuse
 * (wrong-length key, truncated input).
 */
class SecretBoxTest {
    @Test
    void sealOpenRoundTrips() throws Exception {
        byte[] key = SecretBox.keygen();
        byte[] sealed = SecretBox.seal(key, "a message worth protecting".getBytes("UTF-8"));
        assertArrayEquals("a message worth protecting".getBytes("UTF-8"), SecretBox.open(key, sealed));
    }

    @Test
    void sealHandlesEmptyMessage() {
        byte[] key = SecretBox.keygen();
        byte[] sealed = SecretBox.seal(key, new byte[0]);
        assertArrayEquals(new byte[0], SecretBox.open(key, sealed));
    }

    @Test
    void tamperedCiphertextIsRejected() throws Exception {
        byte[] key = SecretBox.keygen();
        byte[] sealed = SecretBox.seal(key, "message".getBytes("UTF-8"));
        sealed[sealed.length - 1] ^= 1;
        assertThrows(DstuException.class, () -> SecretBox.open(key, sealed));
    }

    @Test
    void tamperedNonceIsRejected() throws Exception {
        byte[] key = SecretBox.keygen();
        byte[] sealed = SecretBox.seal(key, "message".getBytes("UTF-8"));
        sealed[0] ^= 1; // first byte of the nonce
        assertThrows(DstuException.class, () -> SecretBox.open(key, sealed));
    }

    @Test
    void wrongKeyIsRejected() throws Exception {
        byte[] key = SecretBox.keygen();
        byte[] otherKey = SecretBox.keygen();
        byte[] sealed = SecretBox.seal(key, "message".getBytes("UTF-8"));
        assertThrows(DstuException.class, () -> SecretBox.open(otherKey, sealed));
    }

    @Test
    void wrongLengthKeyIsRejected() {
        byte[] tooShort = "too short".getBytes();
        assertThrows(IllegalArgumentException.class, () -> SecretBox.seal(tooShort, "message".getBytes()));
    }

    @Test
    void truncatedSealedInputIsRejected() {
        byte[] key = SecretBox.keygen();
        assertThrows(DstuException.class, () -> SecretBox.open(key, "short".getBytes()));
    }
}
